### PR TITLE
fix: add spontaneous request rejection

### PR DESCRIPTION
**Pull request by oggy-dfin**
The IC can currently reject requests at any point in time (e.g., because it's overloaded). But there is no corresponding transition in the spec currently.

### CODE REVIEW

File: spec/index.md [hunk 1]
```diff
@@ -3752,6 +3752,29 @@ The functions `query_as_update` and `system_task_as_update` turns a query functi
...
+S.messages = Older_messages · CallMessage CM · Younger_messages
+(CM.queue = Unordered) or (∀ msg ∈ Older_messages. msg.queue ≠ CM.queue)
+reject_code ∈ { SYS_FATAL, SYS_TRANSIENT, DESTINATION_INVALID }
+```
+
+State after:
+```html
+S.messages =
+    Older_messages ·
+    ResponseMessage {
```
[the remark this answers is not in the document]
Comment (oggy-dfin) [reply spec/index.md]: Adapted now

**Review by mraszyk — APPROVED** [review]
LGTM - thanks!

**Comment by mraszyk** [comment]
I wonder if we could leave the reject code unspecified, too, and thereby close the issue https://github.com/dfinity/interface-spec/issues/171?

**Comment by mraszyk** [comment]
> What's the error returned in #171, `CANISTER_REJECT` or `CANISTER_ERROR`?

It could be both (see [here](https://github.com/dfinity/ic/blob/50733acb057608061c12332a73348aa9308955ef/rs/types/error_types/src/lib.rs#L60)).

> If we leave it unspecified that would allow the system to reject requests to any callee with CANISTER_REJECT, even if the callee wouldn't actually reject the message.

We could say
```
reject_code in {1, 2, 3} or M.callee = ic_principal
```
so that an arbitrary reject code can only be produced by the mgmt canister.

**Comment by oggy-dfin** [comment]
Looking at the errors, it looks like we also want `CANISTER_ERROR` for messages addressed at non-existent canister methods (other than the management canister), but we also seem to be lacking this transition. Given that, I think it's the simplest to just allow all errors, as you initially suggested, WDYT? This will also be an overapproximation of the implementation once we add `SYS_UNKNOWN`, as `SYS_UNKNOWN` will never refund cycles, but nevertheless sounds like a reasonable spec of `SYS_UNKNOWN`.

**Comment by mraszyk** [comment]
> Looking at the errors, it looks like we also want `CANISTER_ERROR` for messages addressed at non-existent canister methods (other than the management canister), but we also seem to be lacking this transition. Given that, I think it's the simplest to just allow all errors, as you initially suggested, WDYT?

We could still rule out `CANISTER_REJECT` for non-mgmt canister and `SYS_UNKNOWN` entirely (given the cycles refund observation). But I wouldn't mind allowing all errors as I initially suggested.

**Comment by oggy-dfin** [comment]
Changed to an arbitrary reject code now. For `SYS_UNKNOWN`, it also future proofs us in case we end up implementing cycle refunds in some cases.

**Comment by oggy-dfin** [comment]
Looking at this one more time, I also allowed out-of-order rejection of call messages now, I think that's also quite sensible.

**Pull request merged**